src/Pulsar.Client/Auth/AuthenticationOauth2.fs

```diff
@@ -0,0 +1,117 @@
+﻿module Pulsar.Client.Auth.Oauth2
+
+open System
+open System.Text.Json
+open System.Net.Http
+open System.Text.Json.Serialization
+open Pulsar.Client.Api
+open Pulsar.Client.Auth.Oauth2Token
+
+type Metadata =
+   {
+        [<JsonPropertyName("issuer")>]
+        Issuer:string
+        
+        [<JsonPropertyName("token_endpoint")>]
+        TokenEndpoint:string
+        
+        [<JsonPropertyName("userinfo_endpoint")>]
+        UserInfoEndpoint:string
+        
+        [<JsonPropertyName("revocation_endpoint")>]
+        RevocationEndpoint:string  
+               
+        [<JsonPropertyName("jwks_uri")>]
+        JwksUri:string
+        
+        [<JsonPropertyName("device_authorization_endpoint")>]
+        DeviceAuthorizationEndpoint:string 
+    }
+type Credentials =
+    {  
+        [<JsonPropertyName("type")>]
+        CredsType : string
+        
+        [<JsonPropertyName("client_id")>]
+        ClientId : string
+        
+        [<JsonPropertyName("client_secret")>]
+        ClientSecret : string 
+
+        [<JsonPropertyName("client_email")>]
+        ClientEmail : string 
+
+        [<JsonPropertyName("issuer_url")>]
+        IssuerUrl : string 
+    }
+
+//Gets a well-known metadata URL for the given OAuth issuer URL.
+//https://tools.ietf.org/id/draft-ietf-oauth-discovery-08.html#ASConfig
+let getWellKnownMetadataUrl (issuerUrl: Uri) : Uri =
+    Uri(issuerUrl.AbsoluteUri + ".well-known/openid-configuration")
+
+let getMetadata (issuerUrl: Uri) : Async<Metadata> =
+    async {
+        use client = new HttpClient()
```

Comment:
Could you make use of HttpClientFactory instead?

Reply:
It would need to add dependency on System.Net.Http and DI like [here](https://docs.microsoft.com/en-us/previous-versions/aspnet/hh995280(v=vs.108)?redirectedfrom=MSDN),  [here](https://www.stevejgordon.co.uk/introduction-to-httpclientfactory-aspnetcore) or [here](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.ihttpclientfactory?view=dotnet-plat-ext-5.0). On the other hand, it uses one request to get metadata and one request to get/renew token; so I'm not sure if pros of using factory will outweigh cons of adding another nuget.

Reply:
Ok, in that case can you pleas pass this instance of httpclient to the TokenClient and reuse it on refresh requests.